[PT2][Quant] Move add/add relu pattern via module partitioner

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102399
* #102398
* __->__ #102397
* #102396
* #102395
* #102394

This diff uses module partitioners to find add and add + relu patterns.

Differential Revision: [D46095330](https://our.internmc.facebook.com/intern/diff/D46095330/)

kimishpatel committed:
[PT2][Quant] Move add/add relu pattern via module partitioner

This diff uses module partitioners to find add and add + relu patterns.

Differential Revision: [D46095330](https://our.internmc.facebook.com/intern/diff/D46095330/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2][Quant] Move add/add relu pattern via module partitioner"


This diff uses module partitioners to find add and add + relu patterns.

Differential Revision: [D46095330](https://our.internmc.facebook.com/intern/diff/D46095330/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PT2][Quant] Move add/add relu pattern via module partitioner"


This diff uses module partitioners to find add and add + relu patterns.

Differential Revision: [D46095330](https://our.internmc.facebook.com/intern/diff/D46095330/)

[ghstack-poisoned]

leslie-fang-intel commented on torch/ao/quantization/_pt2e/graph_utils.py:
```diff
@@ -13,6 +14,7 @@
     {torch.nn.Conv2d, torch.nn.functional.conv2d},
     {torch.nn.ReLU, torch.nn.functional.relu, torch.nn.functional.relu_},
     {torch.nn.BatchNorm2d, torch.nn.functional.batch_norm},
+    {torch.add, operator.add},
```
Thanks for providing this helper function? How about the `operator.iadd`? Should it also been added here?

jerryzh168 replied:
yeah I think this is needed

leslie-fang-intel replied:
Thanks for the reply @jerryzh168 @kimishpatel. BTW, I think current `_EQUIVALENT_TYPES` list has been hardcoded. So PyTorch Extension Libraries such as [intel-extension-for-pytorch](https://github.com/intel/intel-extension-for-pytorch) can't use `Module Partition API` with customized `_EQUIVALENT_TYPES`. Do you think is it reasonable to make it customizable? I have drafted a PR here https://github.com/pytorch/pytorch/pull/102516.

jerryzh168 replied:
yeah I think so, we haven't finalized our APIs though

kimishpatel replied:
> Thanks for providing this helper function? How about the operator.iadd? Should it also been added here?

Yes, adding iadd makes sense.

> BTW, I think current _EQUIVALENT_TYPES list has been hardcoded. So PyTorch Extension Libraries such as [intel-extension-for-pytorch](https://github.com/intel/intel-extension-for-pytorch) can't use Module Partition API with customized _EQUIVALENT_TYPES.

For this can you say more. I briefly looked at the extension but couldnt quite figure what it is doing. Does it introduce custom nn modules for fusion etc? If so I would ask if fusion is something that can be done via graph rewrite? Any other examples might be helpful as well.

Regarding `_EQUIVALENT_TYPES`, yeah it can be something tht is extensible, but I am curious to know your usecase.

kimishpatel committed:
Update on "[PT2][Quant] Move add/add relu pattern via module partitioner"


This diff uses module partitioners to find add and add + relu patterns.

Differential Revision: [D46095330](https://our.internmc.facebook.com/intern/diff/D46095330/)

[ghstack-poisoned]
